### PR TITLE
ARROW-11932: [C++] Implement ArrayBuilder::AppendScalar

### PR DESCRIPTION
**Pull request by lidavidm**
This repurposes the code from ScalarVectorToArray. This doesn't support dictionary scalars.

### CODE REVIEW

File: cpp/src/arrow/array/builder_base.h [hunk 1]
```diff
@@ -116,6 +116,10 @@ class ARROW_EXPORT ArrayBuilder {
   /// This method is useful when appending null values to a parent nested type.
   virtual Status AppendEmptyValues(int64_t length) = 0;
 
+  /// \brief Append a value from a scalar
+  virtual Status AppendScalar(const Scalar& scalar);
+  virtual Status AppendScalars(const ScalarVector& scalars);
```
Comment (pitrou) [inline cpp/src/arrow/array/builder_base.h:121]: I would also suggest the overload `AppendScalar(const Scalar& scalat, int64_t nrepeats)` to append the same scalar a number of times.

Comment (lidavidm) [reply cpp/src/arrow/array/builder_base.h]: Good point, I've added the overload.

File: cpp/src/arrow/array/builder_base.cc [hunk 2]
```diff
@@ -92,6 +95,132 @@ Status ArrayBuilder::Advance(int64_t elements) {
   return null_bitmap_builder_.Advance(elements);
 }
 
+struct AppendScalarImpl {
```
Comment (pitrou) [inline cpp/src/arrow/array/builder_base.cc:98]: Nit, but this should be in the anonymous namespace.

**Comment by pitrou** [comment]
Hmm, it looks like the implementation could be more optimized. When appending multiple scalars (or multiple time the same scalars), the required space could be reserved. Also, at least for primitive types, I would expect appending multiple scalars to be much more streamlined than this.

(it looks like such optimizations could be crafted more easily if `AppendScalar` was a bunch of virtual methods?)

Perhaps create a TODO JIRA for those optimizations, unless you would like to work on them on this PR.

**Comment by lidavidm** [comment]
I changed it to at least call Reserve() since that's simple. I filed ARROW-13197 and also noted the dictionary scalar support there.

**Review by pitrou — APPROVED** [review]

**Comment by lidavidm** [comment]
(Note: Ben left feedback on this in #10511 so I'll address that here.)

**Comment by lidavidm** [comment]
Now fixed to not require allocations/not use a vector as Ben suggested.

File: cpp/src/arrow/array/builder_base.cc [hunk 3]
```diff
@@ -92,6 +95,139 @@ Status ArrayBuilder::Advance(int64_t elements) {
...
+  template <typename T>
+  enable_if_has_string_view<T, Status> Visit(const T&) {
+    return UseBuilder<T>(AppendBuffer{});
+  }
+
+  template <typename T>
+  enable_if_decimal<T, Status> Visit(const T&) {
+    return UseBuilder<T>(AppendValue{});
+  }
+
```
Comment (bkietz) [inline cpp/src/arrow/array/builder_base.cc:160]: +1 for reserving AOT for simple types

```suggestion
  template <typename T>
  enable_if_fixed_width<T, Status> Visit(const T&) {
    auto builder = checked_cast<typename TypeTraits<T>::BuilderType*>(builder_);
    RETURN_NOT_OK(builder->Reserve(n_repeats_ * (scalar_end_ - scalar_begin_)));

    for (int64_t i = 0; i < n_repeats_; i++) {
      for (const std::shared_ptr<Scalar>* raw = scalars_begin_; raw != scalars_end_;
           raw++) {
        auto scalar = checked_cast<const typename TypeTraits<T>::ScalarType*>(raw->get());
        if (scalar->is_valid) {
          builder->UnsafeAppend(scalar->value);
        } else {
          builder->UnsafeAppendNull();
        }
      }
    }
    return Status::OK();
  }

  template <typename T>
  enable_if_base_binary<T, Status> Visit(const T&) {
    int64_t data_size = 0;
    for (const std::shared_ptr<Scalar>* raw = scalars_begin_; raw != scalars_end_;
           raw++) {
      auto scalar = checked_cast<const typename TypeTraits<T>::ScalarType*>(raw->get());
      if (scalar->is_valid) {
        data_size += scalar->value->size();
      }
    }

    auto builder = checked_cast<typename TypeTraits<T>::BuilderType*>(builder_);
    RETURN_NOT_OK(builder->Reserve(n_repeats_ * (scalar_end_ - scalar_begin_)));
    RETURN_NOT_OK(builder->ReserveData(n_repeats_ * data_size));

    for (int64_t i = 0; i < n_repeats_; i++) {
      for (const std::shared_ptr<Scalar>* raw = scalars_begin_; raw != scalars_end_;
           raw++) {
        auto scalar = checked_cast<const typename TypeTraits<T>::ScalarType*>(raw->get());
        if (scalar->is_valid) {
          builder->UnsafeAppend(util::string_view{*scalar->value});
        } else {
          builder->UnsafeAppendNull();
        }
      }
    }
    return Status::OK();
  }
```

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/array/builder_base.cc [hunk 4]
```diff
@@ -92,6 +95,163 @@ Status ArrayBuilder::Advance(int64_t elements) {
   return null_bitmap_builder_.Advance(elements);
 }
 
+namespace {
+struct AppendScalarImpl {
+  template <typename T>
+  enable_if_t<has_c_type<T>::value || is_decimal_type<T>::value, Status> Visit(const T&) {
```
Comment (bkietz) [inline cpp/src/arrow/array/builder_base.cc:101]: This doesn't include `FixedSizeBinaryType`, which doesn't require a ReserveData call

Comment (lidavidm) [reply cpp/src/arrow/array/builder_base.cc:101]: FixedSizeBinaryType didn't work with this specialization anyways, I'll go add the required Builder::Append(Buffer) call that's needed.

**Comment by lidavidm** [comment]
This got included in #10511.

**Pull request closed without merging**